tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,252 @@
+
+        private const string StateMachineNamePrefix = "<";
+        private const string StateMachineNameSuffix = ">";
+        private const string StateMachineFieldsNamePrefix = "<>";
+        private const string ThisName = "__this";
+        private const string StateMachineThisFieldName = StateMachineFieldsNamePrefix + "4" + ThisName;
+        // Should be the same as in debugger_tokens.h managed_profiler_debugger_is_first_entry_field_name
+        private const string StateMachineLiveDebuggerAddedField = "<>dd_liveDebugger_isReEntryToMoveNext";
+        private const BindingFlags AllFieldsBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldInfo[] GetHoistedArgumentsFromStateMachine<TTarget>(TTarget instance, string[] originalMethodParametersName)
+        {
+            if (originalMethodParametersName is null || originalMethodParametersName.Length == 0)
+            {
+                return Array.Empty<FieldInfo>();
+            }
+
+            var foundFields = new FieldInfo[originalMethodParametersName.Length];
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+
+            int found = 0;
+            for (int i = 0; i < allFields.Length; i++)
+            {
+                // we have more fields than parameters, so we break once we found all parameters
+                if (found == foundFields.Length)
+                {
+                    break;
+                }
+
+                var field = allFields[i];
+                // early exit here based on generated name pattern, the assumption will be tested in exploration testing
+                if (field.Name.StartsWith(StateMachineNamePrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                for (var j = 0; j < originalMethodParametersName.Length; j++)
+                {
+                    if (field.Name == originalMethodParametersName[j])
+                    {
+                        foundFields[j] = field;
+                        found++;
+                        break;
+                    }
+                }
+            }
+
+            return foundFields;
+        }
+
+        /// <summary>
+        /// MethodMetadataInfo saves locals from MoveNext's localVarSig,
+        /// This isn't enough in the async scenario because we need to extract more locals that may have been hoisted in the builder object,
+        /// and we need to subtract some locals that exist in the localVarSig but they do not belongs to the kickoff method
+        /// For now, we're capturing all locals that are hoisted (except known generated locals),
+        /// and we're capturing all the locals that appear in localVarSig with ` LogLocal `
```

Comment:
```suggestion
        /// MethodMetadataInfo stores the locals that appear in a method's localVarSig.
        /// This isn't enough in the async scenario (for "MoveNext) because we need to extract more locals that may have been hoisted in the builder object,
        /// and we need to subtract some locals that exist in the localVarSig but they do not belongs to the kickoff method.
        /// For now, we're capturing all locals that are hoisted (except known generated locals),
        /// and we're capturing all the locals that appear in localVarSig with ` LogLocal `
```

Reply:
Can you please clarify what the "locals that exist in the localVarSig but they do not belongs to the kickoff method" are?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26